lib/logger/lib/logger.ex

```diff
@@ -627,41 +628,56 @@ defmodule Logger do
   @doc """
   Logs a message dynamically.
 
-  Use this function only when there is a need to
-  explicitly avoid embedding metadata.
+  Opposite to `log/3`, `debug/2`, `info/2`, and friends, the arguments
+  given to `bare_log/3` are always evaluated. However, you can pass
+  anonymous functions to `bare_log/3` and they will only be evaluated
+  if there is something to be logged.
   """
   @spec bare_log(level, message | (() -> message | {message, keyword}), keyword) ::
           :ok | {:error, :noproc} | {:error, term}
-  def bare_log(level, chardata_or_fun, metadata \\ [])
-      when level in @levels and is_list(metadata) do
+  def bare_log(level, chardata_or_fun, metadata \\ []) do
+    case __maybe_log__(level) do
```

Comment:
I think the name `__maybe_log__` suggests a function that might log or not, but this seems to be fetching the log "context" and checking if it should log or not. So, what do you think about?

```elixir
def bare_log(level, chardata_or_fun, metadata \\ []) do
  case __log_context__(level) do
    :error -> 
      :ok

    log_context ->
     if __should_log__(log_context), do: do_log(log_context, chardata_or_fun, metadata) 
  end
end

@doc false
def __log_context__(level) when level in @levels do
  case __metadata__() do
    {true, pdict} -> {level, Logger.Config.__data__(), pdict}
    {false, _} -> :error
  end
end

@doc false
def __should_log__({level, %{mode: mode, level: min_level}, _pdict}) do
  compare_levels(level, min_level) != :lt and mode != :discard
end
```